### PR TITLE
Dynamodb refactor

### PR DESCRIPTION
**Pull request by jsstevenson**
Various production-oriented additions:
 * Show docs at root endpoint
 * Break up production and dev dependencies
 * Refactor normalizer/database as classes
 * etc

### CODE REVIEW

File: tests/unit/test_chembl_source.py [hunk 1]
```diff
@@ -1,19 +1,22 @@
...
 import pytest
 from therapy.schemas import Drug, MatchType
-from therapy import query
+from therapy.query import Normalizer
 
 
 @pytest.fixture(scope='module')
 def chembl():
-    """Build ChEMBL test fixture."""
+    """Build Wikidata normalizer test fixture."""
```
Comment (korikuzma) [inline tests/unit/test_chembl_source.py:9]:
```suggestion
    """Build ChEMBL normalizer test fixture."""
```

File: tests/unit/test_drugbank_source.py [hunk 2]
```diff
@@ -3,22 +3,26 @@
...
 import pytest
 from therapy.schemas import Drug, MatchType
-from therapy import query
+from therapy.query import Normalizer
 
 
 @pytest.fixture(scope='module')
 def drugbank():
-    """Build DrugBank test fixture."""
+    """Build Wikidata normalizer test fixture."""
```
Comment (korikuzma) [inline tests/unit/test_drugbank_source.py:11]:
```suggestion
    """Build DrugBank normalizer test fixture."""
```

File: tests/unit/test_ncit_source.py [hunk 3]
```diff
@@ -1,19 +1,22 @@
...
 import pytest
 from therapy.schemas import Drug, MatchType
-from therapy import query
+from therapy.query import Normalizer
 
 
 @pytest.fixture(scope='module')
 def ncit():
-    """Build NCIt test fixture"""
+    """Build Wikidata normalizer test fixture."""
```
Comment (korikuzma) [inline tests/unit/test_ncit_source.py:9]:
```suggestion
    """Build NCIt normalizer test fixture."""
```

File: tests/unit/test_therapy_query.py [hunk 4]
```diff
@@ -1,11 +1,27 @@
 """Test the therapy querying method."""
-from therapy.query import normalize, InvalidParameterException
+from therapy.query import Normalizer, InvalidParameterException
 import pytest
 
 
-def test_query():
+@pytest.fixture(scope='module')
+def normalizer():
+    """Build Wikidata normalizer test fixture."""
```
Comment (korikuzma) [inline tests/unit/test_therapy_query.py:8]:
```suggestion
    """Build normalizer test fixture."""
```

File: therapy/database.py [hunk 5]
```diff
@@ -1,28 +1,36 @@
...
         """Initialize Database class."""
-        existing_tables = DYNAMODBCLIENT.list_tables()['TableNames']
-        self.create_therapies_table(DYNAMODB, existing_tables)
-        self.create_meta_data_table(DYNAMODB, existing_tables)
+        if db_url:
+            self.ddb = boto3.resource('dynamodb', region_name=region_name,
+                                      endpoint_url=db_url)
+            self.ddb_client = boto3.resource('dynamodb',
+                                             region_name=region_name,
+                                             endpoint_url=db_url)
```
Comment (korikuzma) [inline therapy/database.py:16]: This never gets called. We should add a flag in cli.py to determine if we are in dev/prod

File: therapy/query.py [hunk 6]
```diff
@@ -1,9 +1,10 @@
 """This module provides methods for handling queries."""
 import re
-from typing import List, Dict, Set
+from typing import List, Dict, Set, Optional
 
 from uvicorn.config import logger
-from therapy.database import THERAPIES_TABLE, METADATA_TABLE, cached_sources
+# from therapy.database import THERAPIES_TABLE, METADATA_TABLE, cached_sources
```
Comment (korikuzma) [inline therapy/query.py:6]:
```suggestion
```

**Review by korikuzma — CHANGES_REQUESTED** [review]
Need to add dev db_url

File: cli.py [hunk 7]
```diff
@@ -20,7 +19,7 @@ class CLI:
     )
     @click.option(
         '--normalizer',
-        help="The normalizer(s) you wish to update separated by spaces."
+        help="The normalizer(s) you wish to update separated by commas."
     )
     def update_normalizer_db(normalizer, all):
```
Comment (korikuzma) [inline cli.py:24]:
```suggestion
    @click.option(
        '--env',
        help="dev or prod"
    )
    def update_normalizer_db(normalizer, all):
```

File: cli.py [hunk 8]
```diff
@@ -31,56 +30,40 @@ def update_normalizer_db(normalizer, all):
...
-                else:
-                    raise Exception("Not a normalizer source.")
+        db: Database = Database()
 
-    def _delete_all_data(self):
-        tables = DYNAMODBCLIENT.list_tables()['TableNames']
-        for table in tables:
-            response = DYNAMODBCLIENT.delete_table(TableName=table)  # noqa F841
-            click.echo(f"Deleted table: {table}")
+        normalizers = normalizer.lower().split(',')
```
Comment (korikuzma) [inline cli.py:35]:
```suggestion
        if env == "prod":
            db: Database = Database()
        elif env == "dev":
            db: Database = Database(db_url='http://localhost:8000')
        else:
            raise Exception('Must enter dev or prod.')
            
        normalizers = normalizer.lower().split(',')
```

File: therapy/query.py [hunk 9]
```diff
@@ -34,349 +34,379 @@ def __init__(self, message):
...
-            exclude all other sources. Case-insensitive. Raises
-            InvalidParameterException if both incl and excl args are provided,
-            or if invalid source names are given.
-        excl: str containing comma-separated names of source to exclude.
-            Will include all other source. Case-insensitive. Raises
-            InvalidParameterException if both incl and excl args are provided,
-            or if invalid source names are given.
+        # TODO testing -- remove when dynamodb implementation complete
+        # sources = {"ncit": "NCIt", "wikidata": "Wikidata",
+        #            "drugbank": "DrugBank"}
```
Comment (korikuzma) [inline therapy/query.py:371]:
```suggestion
```

File: cli.py [hunk 10]
```diff
@@ -4,25 +4,19 @@
...
 
 class CLI:
     """Class for updating the normalizer database via Click"""
 
     @click.command()
-    @click.option(
-        '--all',
-        is_flag=True,
-        help='Update all normalizer databases.'
-    )
```
Comment (korikuzma) [inline cli.py:20]: Why are we removing this flag? We should keep it so that users don't have to type in each normalizer.

Comment (korikuzma) [inline cli.py:20]: Why are we removing this flag? We should keep it so users don't have to write out each normalizer.

File: therapy/database.py [hunk 11]
```diff
@@ -1,28 +1,44 @@
...
+            db_url = environ['THERAPY_NORM_DB_URL']
+            self.ddb = boto3.resource('dynamodb', region_name=region_name,
+                                      endpoint_url=db_url)
+            self.ddb_client = boto3.resource('dynamodb',
+                                             region_name=region_name,
+                                             endpoint_url=db_url)
+        else:
+            self.ddb = boto3.resource('dynamodb', region_name=region_name)
+            self.ddb_client = boto3.resource('dynamodb',
+                                             region_name=region_name)
```
Comment (korikuzma) [inline therapy/database.py:28]:
```suggestion
        if 'THERAPY_NORM_DB_URL' in environ.keys():
            db_url = environ['THERAPY_NORM_DB_URL']
            self.ddb = boto3.resource('dynamodb', region_name=region_name,
                                      endpoint_url=db_url)
            self.ddb_client = boto3.resource('dynamodb',
                                             region_name=region_name,
                                             endpoint_url=db_url)
        else:
            self.ddb = boto3.resource('dynamodb', region_name=region_name)
            self.ddb_client = boto3.resource('dynamodb',
                                             region_name=region_name)
```
The first if statement never gets called.

**Review by korikuzma — CHANGES_REQUESTED** [review]
We need to implement the update_all flag in cli.py. 
Should we be setting the db_url in cli.py or just stick with an env variable?

**Review by jsstevenson — COMMENTED** [review]
👍

**Review by korikuzma — APPROVED** [review]
👍

**Pull request merged**